src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeHandle.cs

```diff
@@ -0,0 +1,258 @@
+
+        public IntPtr DangerousGetHandle() => handle;
+
+        public bool IsClosed => (_state & StateBits.Closed) == StateBits.Closed;
+
+        public abstract bool IsInvalid { get; }
+
+        public void Close() => Dispose();
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            Debug.Assert(_fullyInitialized);
+            InternalRelease(disposeOrFinalizeOperation: true);
+        }
+
+        public void SetHandleAsInvalid()
+        {
+            Debug.Assert(_fullyInitialized);
+
+            // Attempt to set closed state (low order bit of the _state field).
+            // Might have to attempt these repeatedly, if the operation suffers
+            // interference from an AddRef or Release.
+            int oldState, newState;
+            do
+            {
+                oldState = _state;
+                newState = oldState | StateBits.Closed;
+            } while (Interlocked.CompareExchange(ref _state, newState, oldState) != oldState);
+
+            GC.SuppressFinalize(this);
+        }
+
+        protected abstract bool ReleaseHandle();
+
+        public void DangerousAddRef(ref bool success)
+        {
+            // To prevent handle recycling security attacks we must enforce the
+            // following invariant: we cannot successfully AddRef a handle on which
+            // we've committed to the process of releasing.
+
+            // We ensure this by never AddRef'ing a handle that is marked closed and
+            // never marking a handle as closed while the ref count is non-zero. For
+            // this to be thread safe we must perform inspection/updates of the two
+            // values as a single atomic operation. We achieve this by storing them both
+            // in a single aligned int and modifying the entire state via interlocked
+            // compare exchange operations.
+
+            // Additionally we have to deal with the problem of the Dispose operation.
+            // We must assume that this operation is directly exposed to untrusted
+            // callers and that malicious callers will try and use what is basically a
+            // Release call to decrement the ref count to zero and free the handle while
+            // it's still in use (the other way a handle recycling attack can be
+            // mounted). We combat this by allowing only one Dispose to operate against
+            // a given safe handle (which balances the creation operation given that
+            // Dispose suppresses finalization). We record the fact that a Dispose has
+            // been requested in the same state field as the ref count and closed state.
+
+            Debug.Assert(_fullyInitialized);
+            unsafe
+            {
+                // Validate the ref before incrementing the state.
```

Comment:
I do not think that this explicit null check is necessary. We do not protect against null byrefs in C# in general.

Reply:
I went back and forth on that, and figured I'd leave it in to get your opinion.  This was to mimic https://github.com/dotnet/coreclr/blob/2f75b38bd75e88860203a7fad83f5fbadff0b1f4/src/vm/safehandle.cpp#L399-L400.  I'll delete it.

Reply:
The rules for null checks on `this` and `byrefs` are different between managed and unmanaged parts.

You have to go out of your way to get null `this` or `byref` in C# and it you access null `this` or `byref` in C#, you still get nice NullReferenceException.  Thus we do not bother to check explicitly for null for `this` or `byref` in the framework.

However, if you access null `this` or `byref` inside FCall, you will get immediate hard to diagnose failfast, not even NullReferenceException. Thus, we have explicit null checks for `this` or `byref` inside FCalls that try to mimic what you would get out of managed implementation.